skimage/morphology/_rolling_ball_cy.pyx

```diff
@@ -0,0 +1,56 @@
+import numpy as np
+cimport cython
+from libc.math cimport isnan, INFINITY
+
+from .._shared.fused_numerics cimport np_floats
+
+ctypedef np_floats DTYPE_FLOAT
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+def apply_kernel_nan(DTYPE_FLOAT[:,:,:,:] windows,
+                     DTYPE_FLOAT[:,:] kernel,
+                     DTYPE_FLOAT[:,:] cap_height):
+    
+    cdef DTYPE_FLOAT[:, :] out_data = np.zeros((windows.shape[0], windows.shape[1]), dtype=windows.base.dtype)
+    cdef int im_x, im_y, kern_x, kern_y
+    cdef DTYPE_FLOAT min_value, tmp
+
+    with nogil:
+        for im_y in range(windows.shape[0]):
+            for im_x in range(windows.shape[1]):
+                min_value = INFINITY
+                for kern_y in range(kernel.shape[0]):
+                    for kern_x in range(kernel.shape[1]):
+                        tmp = (windows[im_y, im_x, kern_y, kern_x] + cap_height[kern_y, kern_x]) * kernel[kern_y, kern_x]
+                        if min_value > tmp:
+                            min_value = tmp
+                        elif isnan(tmp):
+                            min_value = tmp
+                out_data[im_y, im_x] = min_value
+
+    return out_data.base
+
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+def apply_kernel(DTYPE_FLOAT[:,:,:,:] windows,
+                 DTYPE_FLOAT[:,:] kernel,
+                 DTYPE_FLOAT[:,:] cap_height):
+    
+    cdef DTYPE_FLOAT[:, :] out_data = np.zeros((windows.shape[0], windows.shape[1]), dtype=windows.base.dtype)
+    cdef int im_x, im_y, kern_x, kern_y
+    cdef DTYPE_FLOAT min_value, tmp
+
+    with nogil:
+        for im_y in range(windows.shape[0]):
+            for im_x in range(windows.shape[1]):
+                min_value = INFINITY
+                for kern_y in range(kernel.shape[0]):
+                    for kern_x in range(kernel.shape[1]):
+                        tmp = (windows[im_y, im_x, kern_y, kern_x] + cap_height[kern_y, kern_x]) * kernel[kern_y, kern_x]
+                        if min_value > tmp:
+                            min_value = tmp
```

Comment:
Concerning the treatment of NaNs: if you want can you check how this performs?

```suggestion
                        if not min_value < tmp:
                            min_value = tmp
                            if isnan(min_value):
                                break
                    if isnan(min_value)
                        break
                        
```
Not sure if I got the indents right, but the idea is that once we've found a NaN the following values don't matter anymore and we can break out of the two innermost loops early. `not min_value < tmp` will be True for NaNs as well as for smaller values. The reason the `break` statement is guarded by the `if not` statement is that I expect it to be False more often than True -> we find a new `min_value` less often than we reject the current `tmp` value; is that reasonable?

Even if this is reasonably fast, we might still go for your suggestion with two kernels. It's a great idea and the code of each kernel is straight forward and probably easier to maintain longterm than my suggestion. And we can still use the idea of breaking out of the 2 inner loops early once we find a NaN!

Nevertheless it's still fun attempting to crack this problem with one solution. :grin:

Reply:
I thought about breaking out of loops, too. My hunch is that it will suffer the same fate as only looping over elements for which `kernel[ker_y, ker_x] == 1` (i.e., only the ones that are inside the spacial ellipsis). It seems like avoiding work, because it saves the multiplication as well as a lot of iterations, but for some reason turns out about 3x slower than just blindly iterating the entire kernel. I don't know why though; perhaps this can be improved? I will give the breaking for `NaN` a shot and do the timings. 

Testing `min_value<tmp` or `min_value>tmp` will make no difference for branch prediction I suspect. We are iterating the image left->right and top->bottom, and the image can have a gradient going light->dark or dark->light or something different entirely. Without knowing, both guesses should be about equally good.

Reply:
I did some benchmarking, and weirdly enough something like 

```
if not min_value <= tmp and not isnan(min_value):
    min_value = tmp
```
is faster than
```
if min_value > tmp:
    min_value = tmp
elif isnan(tmp):
    min_value = tmp
```
Not sure what the compiler is doing here ...

Either way, the fastest version seems to be to use `break` if a `NaN` is encountered, but to still keep a nan-free kernel in parallel (2x faster for NaN-free images).

Here are my benchmark results (https://github.com/FirefoxMetzger/scikit-image/commit/af47c2437a6afc1da185a90dfb2c4ef4ceaddfad):

```
[ 16.67%] ··· benchmark_morphology.RollingBall.peakmem_rollingball                                                   ok
[ 16.67%] ··· ======== =======
               radius
              -------- -------
                 25     99.3M
                 50      101M
                 75      101M
                100      102M
                150      105M
                200      108M
              ======== =======

[ 33.33%] ··· benchmark_morphology.RollingBall.time_rollingball                                                      ok
[ 33.33%] ··· ======== =========
               radius
              -------- ---------
                 25     452±0ms
                 50     1.74±0s
                 75     3.62±0s
                100     6.20±0s
                150     13.7±0s
                200     26.2±0s
              ======== =========

[ 50.00%] ··· benchmark_morphology.RollingBall.time_rollingball_nan                                                  ok
[ 50.00%] ··· ======== =========
               radius
              -------- ---------
                 25     837±0ms
                 50     2.64±0s
                 75     4.35±0s
                100     5.98±0s
                150     9.45±0s
                200     14.5±0s
              ======== =========
```